### DISCUSSION TITLE
Build order issue using a cmake subproject dependency

### DISCUSSION THREAD

**Discussion by EmmanuelP** [mesonbuild/meson#9935] [Q&A]
Hi,

Before reporting it as an issue in meson, I'd like to know if I'm doing it wrong.

I have a pretty simple repository with open62541 dependency as a subproject. open62541 build system is based on cmake. The meson part is the following:

``` meson
open62541 = cmake.subproject('open62541', options:open62541_options)
opc_dep = open62541.dependency('open62541')

opctest_dependencies = [opc_dep]

apps = ['producer', 'consumer']

foreach a: apps
  executable(a, a + '.c', dependencies: opctest_dependencies, install: true)
endforeach
```

The problem is when I invoke ninja, the two executables are always built before the open62541 subproject. Is there something wrong in the way I have declared the dependencies ?

The full project is here: [https://github.com/EmmanuelP/opctest](https://github.com/EmmanuelP/opctest)

**Comment by andy5995** [comment]
@EmmanuelP Looking at the [relevant docs](https://mesonbuild.com/CMake-module.html#cmake-options-object), it appears you'll need to make some changes.

Also, could you edit your post to correct the link to your GitHub repo? It's linking to the discussion you created.

**Reply by EmmanuelP** [reply]
> @EmmanuelP Looking at the [relevant docs](https://mesonbuild.com/CMake-module.html#cmake-options-object), it appears you'll need to make some changes.

Would you mind to elaborate ? I'm not seeing anything obvious in the documentation.

> Also, could you edit your post to correct the link to your GitHub repo? It's linking to the discussion you created.

Done.

**Reply by andy5995** [reply]
> > @EmmanuelP Looking at the [relevant docs](https://mesonbuild.com/CMake-module.html#cmake-options-object), it appears you'll need to make some changes.
> 
> Would you mind to elaborate ? I'm not seeing anything obvious in the documentation.
> 

I mean to link to [this section](https://mesonbuild.com/CMake-module.html#configuration-options) on that page. But when I look at the meson.build in your repo, it looks fine (a bit more complete than the code sample you posted above).

When I try to build your project, I get an error pretty quick because an include file can't be found:

```
[andy@prometheus buildddir]$ ninja
[1/84] Compiling C object producer.p/producer.c.o
FAILED: producer.p/producer.c.o 
cc -Iproducer.p -I. -I.. -Isubprojects/open62541/__CMake_build/src_generated -I../subprojects/open62541/__CMake_build/src_generated -I../subprojects/open62541/arch -I../subprojects/open62541/deps -I../subprojects/open62541/include -I../subprojects/open62541/plugins/include -I../subprojects/open62541/src -I../subprojects/open62541/src/pubsub -Isubprojects/open62541/__CMake_build -I../subprojects/open62541/__CMake_build -Isubprojects/open62541 -I../subprojects/open62541 -I../subprojects/open62541/src/client -I../subprojects/open62541/plugins -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c99 -g -DUA_ARCHITECTURE_POSIX -MD -MQ producer.p/producer.c.o -MF producer.p/producer.c.o.d -o producer.p/producer.c.o -c ../producer.c
In file included from ../subprojects/open62541/include/open62541/util.h:11,
                 from ../subprojects/open62541/include/open62541/server.h:20,
                 from ../producer.c:2:
../subprojects/open62541/include/open62541/types.h:21:10: fatal error: open62541/statuscodes.h: No such file or directory
   21 | #include <open62541/statuscodes.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[2/84] Generating subprojects/open62541/custom_tgt_2 with a custom command
Unable to find generated file. This can cause the build to fail:
/home/andy/src/opctest/buildddir/subprojects/open62541/__CMake_build/PRE_BUILD
[3/84] Generating subprojects/open62541/custom_tgt_0 with a custom command
Unable to find generated file. This can cause the build to fail:
/home/andy/src/opctest/buildddir/subprojects/open62541/__CMake_build/PRE_BUILD
[4/84] Generating subprojects/open62541/custom_tgt_3 with a custom command
Unable to find generated file. This can cause the build to fail:
/home/andy/src/opctest/buildddir/subprojects/open62541/__CMake_build/PRE_BUILD
[5/84] Generating subprojects/open62541/custom_tgt_1 with a custom command
Unable to find generated file. This can cause the build to fail:
/home/andy/src/opctest/buildddir/subprojects/open62541/__CMake_build/PRE_BUILD
[6/84] Generating subprojects/open62541/custom_tgt_4 with a custom command
INFO:__main__:Preprocessing /home/andy/src/opctest/subprojects/open62541/tools/schema/Opc.Ua.NodeSet2.Reduced.xml
INFO:__main__:Preprocessing /home/andy/src/opctest/subprojects/open62541/tools/schema/Opc.Ua.NodeSet2.EventsMinimal.xml
INFO:__main__:Preprocessing /home/andy/src/opctest/subprojects/open62541/tools/schema/Opc.Ua.NodeSet2.Part8_Subset.xml
INFO:__main__:Generating Code for Backend: open62541
INFO:__main__:NodeSet generation code successfully printed
Unable to find generated file. This can cause the build to fail:
/home/andy/src/opctest/buildddir/subprojects/open62541/__CMake_build/PRE_BUILD
ninja: build stopped: subcommand failed.
```

But when I cd to `/home/andy/src/opctest/buildddir/subprojects/open62541/__CMake_build` and run `ninja`, the library builds, and if I cd back to the builddir root, and type ninja, the executable builds.

Anyway, I added the 3rd line and changed the 4th a little, and got the build to work the way you wanted (dependency first, exe 2nd).

```
open62541 = cmake.subproject('open62541', options:open62541_options)
opc_dep = open62541.dependency('open62541')
dep_open62541 = dependency('open62541', fallback : ['open62541', 'open62541_dep'])
opctest_dependencies = [dep_open62541]
```
The `open62541_dep` variable is from the meson.build created in builddir/subprojects/open62541 during setup. You can see that, as well as the fallback keyword, explained more [here](https://mesonbuild.com/Subprojects.html).

There may be a cleaner way to do it, but as I've been working on it for 2 hours, it's time for me to quit the PC!

**Reply by EmmanuelP** [reply]
Thanks a lot Andy for the time you spent on my issue.

> When I try to build your project, I get an error pretty quick because an include file can't be found:

That is because the `producer` executable is built before the open62541 subproject. On the second invocation of ninja, it works.

**Reply by andy5995** [reply]
Did you try using the 'fallback' option as I suggested?

**Reply by EmmanuelP** [reply]
Oh sorry, I did not tell.

Yes it works.

**Reply by EmmanuelP** [reply]
Oh, I have just rechecked, and in the end your suggestion does not work. I did not notice the compilation step of producer and consumer.

![image](https://user-images.githubusercontent.com/136864/152748091-5572f8a4-ba29-462e-96ca-e5fe45a41089.png)

See how producer and consumer are compiled before open62541 (or at the same time ?).

**Reply by andy5995** [reply]
@EmmanuelP , And when I ran `ninja clean` and `ninja`, it won't rebuild. Hmm... Probably now would be a good time to open a ticket (There's a link on the right to "create issue from discussion"). I'm not a meson expert, but comparing your meson.build to the meson docs, I don't see a problem.

I wonder if it's a newly introduced bug. @wbthomason, has [cpzlib](https://github.com/wbthomason/cpzlib) ever built successfully with your current meson setup?

**Reply by andy5995** [reply]
> Probably now would be a good time to open a ticket (There's a link on the right to "create issue from discussion")

@EmmanuelP Or maybe it's related to existing tickets: https://github.com/mesonbuild/meson/issues/8089#issuecomment-743277109 or https://github.com/mesonbuild/meson/issues/9062

**Reply by wbthomason** [reply]
@andy5995 It does not build with my current setup, but it was building the last time I tried (almost a year ago; that project is currently semi-dead).

**Reply by wbthomason** [reply]
Slight correction: once I update the version of `celero`, it does in fact build as-is (with the invocation `CXX=clang++ meson build && meson compile -C build`. (g++ may also work, but I haven't tried yet)

**Reply by andy5995** [reply]
@wbthomason This is the output I get using the current state of your git repo:

```
[andy@prometheus cpzlib]$ CXX=clang++ meson build && meson compile -C build
The Meson build system
Version: 0.61.1
Source dir: /home/andy/src/cpzlib
Build dir: /home/andy/src/cpzlib/build
Build type: native build
Project name: cpzlib
Project version: 0.1
C++ compiler for the host machine: clang++ (clang 13.0.0 "clang version 13.0.0")
C++ linker for the host machine: clang++ ld.bfd 2.36.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /home/linuxbrew/.linuxbrew/bin/pkg-config (0.29.2)
Found CMake: /usr/bin/cmake (3.22.2)
Run-time dependency eigen3 found: NO (tried pkgconfig and cmake)
Looking for a fallback subproject for the dependency eigen3

Executing subproject eigen 

eigen| Project name: eigen
eigen| Project version: 3.3.9
eigen| C++ compiler for the host machine: clang++ (clang 13.0.0 "clang version 13.0.0")
eigen| C++ linker for the host machine: clang++ ld.bfd 2.36.1
eigen| Build targets in project: 0
eigen| Subproject eigen finished.

WARNING: Variable 'eigen3' in the subproject 'subprojects/eigen-3.3.9' is not found
Dependency eigen3 from subproject subprojects/eigen-3.3.9 found: NO

meson.build:7:0: ERROR: Dependency 'eigen3' is required but not found.
```

**Reply by wbthomason** [reply]
@andy5995 I cannot reproduce that (I just tried a fresh clone and build, which works as expected). Perhaps this is because I have Eigen installed on my system already, and do not need to rely on the subproject?

**Reply by wbthomason** [reply]
I'll note that I have another (currently private) project also successfully using a CMake subproject dependency; the relevant part of my `meson.build` is:
```meson
# Dependencies
cmake = import('cmake')

autodiff = dependency('autodiff', required: false)
if not autodiff.found()
  autodiff_cmake_opts = cmake.subproject_options()
  autodiff_cmake_opts.add_cmake_defines({'AUTODIFF_BUILD_EXAMPLES': false})
  autodiff_cmake_opts.set_override_option('cpp_std', 'c++20')
  autodiff_cmake = cmake.subproject('autodiff', options: autodiff_cmake_opts)
  autodiff = autodiff_cmake.dependency('autodiff')
endif
```

**Reply by andy5995** [reply]
> @andy5995 I cannot reproduce that (I just tried a fresh clone and build, which works as expected). Perhaps this is because I have Eigen installed on my system already, and do not need to rely on the subproject?

@wbthomason Just wanted to let you know that after I installed eigen (on Manjaro), cpzlib builds (and using g++ 11.1.0).

**Reply by wbthomason** [reply]
Thanks - there must be a problem with my eigen subproject setup.

**Comment by andy5995** [comment]
@eli-schwartz @jpakkane I referred to [this meson.build](https://github.com/wbthomason/cpzlib/blob/d0361c91d634480e60988a396c02334b5fccb44e/meson.build#L16-L20) a little. Isn't the dependency supposed to build just by using those two lines within the 'if' condition there? I couldn't get this to work without using the `fallback` keyword.

**Reply by eli-schwartz** [reply]
The `fallback` kwarg is not the same as an explicit `subproject()` invocation. Also automatic fallback assumes the wrapped subproject uses meson, not cmake.
